src/MessagePack.GeneratorCore/MessagePackCompilation.cs

```diff
@@ -0,0 +1,316 @@
+               .Select(x => MetadataReference.CreateFromFile(x))
+               .ToList();
+
+            var sources = new List<string>();
+            var locations = new List<string>();
+            foreach (var csproj in csprojs)
+            {
+                CollectDocument(csproj, sources, locations);
+            }
+
+            var hasAnnotations = false;
+            foreach (var file in sources.Distinct())
+            {
+                var text = File.ReadAllText(file.Replace('\\', Path.DirectorySeparatorChar), Encoding.UTF8);
+                var syntax = CSharpSyntaxTree.ParseText(text, parseOption);
+                syntaxTrees.Add(syntax);
+                if (Path.GetFileNameWithoutExtension(file) == "Attributes")
+                {
+                    var root = await syntax.GetRootAsync(cancellationToken);
+                    if (root.DescendantNodes().OfType<ClassDeclarationSyntax>().Any(x => x.Identifier.Text == "MessagePackObjectAttribute"))
+                    {
+                        hasAnnotations = true;
+                    }
+                }
+            }
+            if (!hasAnnotations)
+            {
+                syntaxTrees.Add(CSharpSyntaxTree.ParseText(DummyAnnotation, parseOption));
+            }
+
+            var nazoloc = locations.Distinct();
+            foreach (var item in locations.Distinct().Where(x => !x.Contains("MonoBleedingEdge")))
+            {
+                metadata.Add(MetadataReference.CreateFromFile(item));
+            }
+
+            var compilation = CSharpCompilation.Create(
+                "MessagepackCodeGenTemp",
+                syntaxTrees,
+                metadata,
+                new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary, allowUnsafe: true));
+
+            return compilation;
+        }
+
+        static void CollectDocument(string csproj, List<string> source, List<string> metadataLocations)
+        {
+            XDocument document;
+            using (var sr = new StreamReader(csproj, true))
+            {
+                var reader = new XmlTextReader(sr);
+                reader.Namespaces = false;
+
+                document = XDocument.Load(reader, LoadOptions.None);
+            }
+
+            var csProjRoot = Path.GetDirectoryName(csproj);
+            var framworkRoot = Path.GetDirectoryName(typeof(object).Assembly.Location);
+
+            // Legacy
+            // <Project ToolsVersion=...>
+            // New
+            // <Project Sdk="Microsoft.NET.Sdk">
+
+            var proj = document.Element("Project");
+            var legacyFormat = proj.Attribute("Sdk")?.Value != "Microsoft.NET.Sdk";
+
+            if (!legacyFormat)
+            {
+                // TODO:parase compile remove
+                foreach (var file in IterateCsFileWithoutBinObj(Path.GetDirectoryName(csproj)))
+                {
+                    source.Add(file);
+                }
+                // TODO:resolve NuGet reference
```

Comment:
For our use case this is a blocker, but I understand that it isn't trivial to implement.

Reply:
Yes, I should support it but it will be a later.
Old mpc has many issues so must move to this model in v2.